File: cracking_the_coding_interview/2-5.cpp
#include <iostream>
#include <vector>
using namespace std;

// 15分45秒
struct Node{
    int val;
    Node* next;
};

Node* sumList(Node* node1, Node* node2, int advance){
    if(!(node1 && node2)){
        return nullptr;
    }

    int num1 = node1 ? node1->val : 0;
    int num2 = node2 ? node2->val : 0;
    int sum = num1 + num2 + advance;
    advance = sum / 10;
    sum %= 10;
    Node* current = new Node;
    current->val = sum;
    node1 = node1 ? node1->next : node1;
    node2 = node2 ? node2->next : node2;
    current->next = sumList(node1, node2, advance);
    return current;
}

Node* createNextNode(const vector<int>& nums, size_t pos){
    if(pos>= nums.size()){
        return nullptr;
    }
    Node* current = new Node;
    current->val = nums[pos];
    current->next = createNextNode(nums, pos+1);
    return current;
}

void print(Node* node){
    while(node){
        cout << node->val;
        if(node->next == nullptr){
            break;
        }
        node = node->next;
    }
    cout << endl;
}

int main(void){
    vector<int> nums1 = {7, 1, 6};
    vector<int> nums2 = {5, 9, 2};
    Node* node1 = createNextNode(nums1, 0);
    Node* node2 = createNextNode(nums2, 0);
    print(node1);
    print(node2);
    Node* ans = sumList(node1, node2, 0);
    print(ans);
    return 0;
}

File: cracking_the_coding_interview/2-5.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def add_list(node1, node2):
    return compute(node1, node2, 0)

def compute(node1, node2, carry):
    if not node1 and not node2:
        return None
    
    v1 = 0
    v2 = 0
    if node1:
        v1 = node1.val
        node1 = node1.next
    if node2:
        v2 = node2.val
        node2 = node2.next
    
    ans = v1 + v2 + carry
    carry = ans // 10
    ans %= 10
    node = ListNode(ans)
    node.next = compute(node1, node2, carry)
    return node

def create_linked_list(nums, i):
    if i >= len(nums):
        return None
    node = ListNode(nums[i])
    node.next = create_linked_list(nums, i+1)
    return node

def output(node):
    if not node:
        return
    
    print(node.val)
    output(node.next)

if __name__ == '__main__':
    nums1 = [7, 1, 6]
    nums2 = [5, 9, 2]
    head1 = create_linked_list(nums1, 0)
    head2 = create_linked_list(nums2, 0)
    res = add_list(head1, head2)
    output(res)